### PR TITLE
fix(flex-container): make action buttons slot work nicely in `limel-dialog`

### PR DESCRIPTION
**Pull request by Kiarokh**
buttons are located in the footer slot.
fix https://github.com/Lundalogik/crm-feature/issues/2232

## Review:
- [x] Commits are [atomic](https://seesparkbox.com/foundry/atomic_commits_with_git)
- [x] Commits have the correct *type* for the changes made
- [x] Commits with *breaking changes* are marked as such

### Browsers tested:
(Check any that applies, it's ok to leave boxes unchecked if testing something didn't seem relevant.)

Windows:
- [ ] Chrome
- [ ] Edge
- [ ] Firefox

Linux:
- [ ] Chrome
- [ ] Firefox

macOS:
- [x] Chrome
- [ ] Firefox
- [ ] Safari

Mobile:
- [ ] Chrome on Android
- [ ] iOS

### CODE REVIEW

File: src/components/dialog/examples/dialog-form.tsx [hunk 1]
```diff
@@ -75,31 +75,27 @@ export class DialogFormExample {
...
-                <limel-flex-container slot="button" reverse={true}>
+                <limel-flex-container slot="button">
+                    <limel-button label="Cancel" onClick={this.closeDialog} />
                     <limel-button
                         primary={true}
                         label="Save"
                         disabled={!this.nameValid() || !this.ageValid()}
                         onClick={this.submitForm}
                     />
-                    <limel-button label="Cancel" onClick={this.closeDialog} />
```
Comment (adrianschmidt) [inline src/components/dialog/examples/dialog-form.tsx:86]: > Having this was completely unnecessary from the beginning.
> Not sure why it was added in the example?
> Maybe to showcase what `flex-container` can do?!
> But consumers copy/paste this exact code
> and it potentially makes troubles for us.
> For example when trying to add fixes such as the previous commit.
> Which will cause the buttons to jump to the left side of the dialog,
> if consumer has copy/pasted `reverse={true}` as well.

The idea was that it would be easier for the users if a single tab-press brought them to the primary button, instead of having to tab past the cancel-button first.

It was probably misguided… 🤷‍♂️

File: src/components/dialog/examples/dialog-closing-actions.tsx [hunk 2]
```diff
@@ -40,15 +34,15 @@ export class DialogClosingActionsExample {
         ];
     }
 
-    private triggerOnClick() {
+    private triggerOnClick = () => {
```
Comment (adrianschmidt) [inline src/components/dialog/examples/dialog-closing-actions.tsx:69]: ❤️

I've been thinking for a while I should go through and make these changes. Thank you! 🙏

Comment (adrianschmidt) [reply src/components/dialog/examples/dialog-closing-actions.tsx:69]: I'm not sure about the commit message though? "improve ui for config options in examples"?

Am I just being daft again? 😆

File: src/components/dialog/examples/dialog-form.tsx (hunk 1 above)
Comment (Kiarokh) [reply src/components/dialog/examples/dialog-form.tsx:86]: I see... I'm a bad example and I no clue about how people navigate with keyboard. But I think their expectation is to navigate from left to right. So I'd suggest we skip this, unless we have some usability/accessibility guidelines that says this is the expected behaviour.

File: src/components/dialog/examples/dialog-closing-actions.tsx (hunk 2 above)
Comment (Kiarokh) [reply src/components/dialog/examples/dialog-closing-actions.tsx:69]: 😄  I don't know dude. I just copied one of your own commit messages for an exact change in another component 😅 (don't remember  where it was, but I guess it was when we fixed `readonly` styles for the input field.

Comment (Kiarokh) [reply src/components/dialog/examples/dialog-closing-actions.tsx:69]: but feel free to change. I have no clue what "improve ui for config options in examples" means 😄

File: src/components/flex-container/flex-container.scss [hunk 3]
```diff
@@ -30,7 +28,6 @@
 :host(limel-flex-container[align='stretch']) {
     align-items: stretch;
 }
-
```
Comment (adrianschmidt) [inline src/components/flex-container/flex-container.scss:33]: I've split this code formatting change into a separate commit from the functional change.

File: src/components/dialog/examples/dialog-form.tsx (hunk 1 above)
Comment (adrianschmidt) [reply src/components/dialog/examples/dialog-form.tsx:86]: Yeah, I agree. I don't have any guidelines for it, just an idea I had before I knew better 😆

Kill it with fire! 🙂

File: src/components/dialog/examples/dialog-closing-actions.tsx (hunk 2 above)
Comment (adrianschmidt) [reply src/components/dialog/examples/dialog-closing-actions.tsx:69]: Ah! Ok, I found the commit. The change that I was referring to there wasn't the change from `function` to `=>`, it was the fact that I had changed from using limel-switch and limel-button in a few different examples, to using limel-checkbox, for consistency…

I should have split that commit into two separate ones I guess 😄

Comment (adrianschmidt) [reply src/components/dialog/examples/dialog-closing-actions.tsx:69]: I'll come up with some better commit message. Tomorrow. It's time to log off for today 😅

Comment (adrianschmidt) [reply src/components/dialog/examples/dialog-closing-actions.tsx:69]: ⚡

File: src/components/flex-container/flex-container.scss [hunk 4]
```diff
@@ -49,3 +46,14 @@
...
+// Since consumers often copy/paste from examples of `limel-dialog` from the
+// documentations, they often put their buttons in a `limel-flex-container`
+// which is used in the footer slot of those examples with buttons.
+// So to prevent unwanted layout problems of buttons having no distance, we use
+// below code to target `limel-flex-container`s in those slots.
+:host(limel-flex-container[slot='button']) {
+    gap: 0.5rem; // Makes sure buttons get some default distance
+    width: 100%;
+    justify-content: flex-end;
+}
```
Comment (adrianschmidt) [inline src/components/flex-container/flex-container.scss:59]: Hmm… does this really have to be added in `flex-container.scss`? Since it's added on `:host`, I think it could be added by the dialog's css, and I wonder if that wouldn't make more sense?

Also, I've been thinking about whether we really need to specify that it should be limel-flex-container at all… 🤔 Maybe we could just say in the dialog css, that whatever element has the `slot="button"` attribute should have `display: flex` and these additional styles?

Comment (adrianschmidt) [inline src/components/flex-container/flex-container.scss:59]: @jgroth I dug through the history to find where limel-flex-container was added, and it was added way back in [v17.0.0](https://github.com/Lundalogik/lime-elements/releases/tag/v17.0.0), without reference to any issue, or anything else really explaining what the purpose was. It was added in a commit by you, so I'm wondering if you remember any details? 🙂

File: src/components/dialog/examples/dialog-closing-actions.tsx [hunk 5]
```diff
@@ -2,7 +2,36 @@ import { Component, h, State } from '@stencil/core';
...
+ * sighted users. Because majority of them users know that clicking or tapping
+ * outside the dialog closes it.
+ *
+ * Such buttons usually are labeled as `OK`, `Dismiss` or `Close`.
+ * :::tip
+ * When to use action buttons for simple "close" actions?
+ * - In fullscreen dialogs where clicking outside to close is hard.
+ * - When big dialogs are opened on phones, which make tapping outside hard for users.
+ * - When designing with accessibility in mind, and for those users who
+ * use screen readers to navigate the user interface.
```
Comment (adrianschmidt) [inline src/components/dialog/examples/dialog-closing-actions.tsx:19]: For accessibility, it feels like there should always be a close button, but it could be visually hidden… but finding the best solution for that is perhaps more work than we're willing to put in right now?

File: src/components/flex-container/flex-container.scss (hunk 4 above)
Comment (jgroth) [reply src/components/flex-container/flex-container.scss:59]: Haha no sorry I don't remember :smile: I think we just wanted some element for the consultants that they could use if they did not know a lot of CSS

Comment (adrianschmidt) [reply src/components/flex-container/flex-container.scss:59]: Yeah… and they probably like it! But I don't… 😆

Or, well, I'm just worried it will be causing us trouble… I feel like we should at least try not to build ourselves further into it by making it more or less required in dialog footers… 🤔

File: src/components/dialog/examples/dialog-closing-actions.tsx [hunk 6]
```diff
@@ -2,7 +2,36 @@ import { Component, h, State } from '@stencil/core';
...
  * Custom closing actions
+ *
+ * Action buttons in dialogs can be used to add a clear visual indication for
+ * the sighted users to realize that the dialog can be closed by pressing
+ * a button as well.
+ * This may sometimes be considered an unnecessary usage of action buttons for
+ * sighted users. Because majority of them users know that clicking or tapping
+ * outside the dialog closes it.
+ *
+ * Such buttons usually are labeled as `OK`, `Dismiss` or `Close`.
```
Comment (adrianschmidt) [inline src/components/dialog/examples/dialog-closing-actions.tsx:13]: Backticks formats the text as code. That _looks_ good in this case, but I don't think it's semantically good… I think quote marks, cursive text, and/or bold text would serve the purpose better.

```suggestion
 * Such buttons are usually labeled ***OK***, ***Dismiss*** or ***Close***.
```

File: src/components/dialog/examples/dialog-closing-actions.tsx [hunk 7]
```diff
@@ -2,7 +2,36 @@ import { Component, h, State } from '@stencil/core';
...
+ * dismissing the dialog.
+ * :::tip
+ * When to use custom closing actions?
+ * - To make sure that the user really reads and understands the dialog's content.
+ * - To make sure that the user does not accidentally click outside and close the dialog.
+ * :::
+ * For such cases, avoid generic labels like `ok`, or `close` which unconciously
+ * motivate users to dismiss the message; and instead use more purposeful labels
+ * such as `I understand`, `Looks good!`, `Continue`, and similar;
+ * like in the example below.
```
Comment (adrianschmidt) [inline src/components/dialog/examples/dialog-closing-actions.tsx:32]:
```suggestion
 * For such cases, avoid generic labels like ***OK***, or ***Close*** which unconsciously
 * motivate users to dismiss the message; and instead use more purposeful labels
 * such as ***I understand***, ***Looks good!***, ***Continue***, and similar;
 * like in the example below.
```

File: src/components/dialog/examples/dialog-closing-actions.tsx [hunk 8]
```diff
@@ -34,21 +57,24 @@ export class DialogClosingActionsExample {
                     the escape key. Only the Ok-button triggers a close event.
```
Comment (adrianschmidt) [inline src/components/dialog/examples/dialog-closing-actions.tsx:57]: This still refers to the "Ok-button", but the button has been relabelled.

File: src/components/dialog/examples/dialog-closing-actions.tsx [hunk 9]
```diff
@@ -2,7 +2,36 @@ import { Component, h, State } from '@stencil/core';
 
 /**
  * Custom closing actions
+ *
+ * Action buttons in dialogs can be used to add a clear visual indication for
+ * the sighted users to realize that the dialog can be closed by pressing
+ * a button as well.
+ * This may sometimes be considered an unnecessary usage of action buttons for
+ * sighted users. Because majority of them users know that clicking or tapping
+ * outside the dialog closes it.
```
Comment (adrianschmidt) [inline src/components/dialog/examples/dialog-closing-actions.tsx:11]: I think this section is a little incoherent

> … can be closed by pressing a button as well.

"As well" as what? It sounds like I'm meant to have read some missing section before this.

The whole paragraph seems to assume that the reader, and all users, see clicking outside a dialog as the _primary_ way to close it, and that having a close-button is some new invention, the purpose of which needs explanation. At least to my ears, that makes this paragraph sound quite weird.

File: src/components/flex-container/flex-container.scss (hunk 4 above)
Comment (adrianschmidt) [reply src/components/flex-container/flex-container.scss:59]: Hmm… no… it doesn't seem like it can be added from the dialog's styling… 😞

But I did find another solution, which I've added as a separate commit.

File: src/components/dialog/examples/dialog-closing-actions.tsx (hunk 8 above)
Comment (adrianschmidt) [reply src/components/dialog/examples/dialog-closing-actions.tsx]: ⚡

File: src/components/dialog/examples/dialog-closing-actions.tsx (hunk 9 above)
Comment (adrianschmidt) [reply src/components/dialog/examples/dialog-closing-actions.tsx:11]: @Kiarokh I'd like this paragraph to get a touch-up, but apart from that I'm ready to approve.

**Review by adrianschmidt — CHANGES_REQUESTED** [review]

File: src/components/dialog/examples/dialog-closing-actions.tsx (hunk 5 above)
Comment (Kiarokh) [reply src/components/dialog/examples/dialog-closing-actions.tsx:20]: I think so… but that's a good idea to add an invisible button

File: src/components/dialog/examples/dialog-closing-actions.tsx (hunk 9 above)
Comment (Kiarokh) [reply src/components/dialog/examples/dialog-closing-actions.tsx:11]: Good point! What about this: 

to realize that the dialog can be closed by pressing a button as well as the conventional closing way, which is clicking outside the dialog.

Comment (adrianschmidt) [reply src/components/dialog/examples/dialog-closing-actions.tsx:11]: Yeah, that's better at least. It's still a little weird to me to consider clicking outside the dialog the _main_ way of dismissing it, but sure 👍

Comment (adrianschmidt) [reply src/components/dialog/examples/dialog-closing-actions.tsx:11]: Will you add a fixup @Kiarokh? I'd love to get this merged 🙂

Comment (adrianschmidt) [reply src/components/dialog/examples/dialog-closing-actions.tsx:11]: I'm going to approve and merge this. You can always update the text later 🙂

**Review by adrianschmidt — DISMISSED** [review]

File: src/components/flex-container/flex-container.scss [hunk 10]
```diff
@@ -49,3 +46,17 @@
...
+:host(limel-flex-container[slot='button'][direction='horizontal'][reverse]) {
+    justify-content: flex-start;
```
Comment (adrianschmidt) [inline src/components/flex-container/flex-container.scss:61]: I added this so that we don't "break" existing solutions that use `reverse={true}`.

Now:
image

Before:
image

**Review by adrianschmidt — APPROVED** [review]

File: src/components/flex-container/flex-container.scss (hunk 10 above)
Comment (Kiarokh) [reply src/components/flex-container/flex-container.scss:61]: smart

File: src/components/flex-container/flex-container.scss (hunk 4 above)
Comment (Kiarokh) [reply src/components/flex-container/flex-container.scss:59]: I really like your solution. Thanks for tweaking my commits!
